src/rrd/mod.rs

```diff
@@ -0,0 +1,529 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use core::time;
+use sqlx::{self, Sqlite};
+use std::{io::Write, path::Path};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("invalid window, can't be zero")]
+    InvalidWindow,
+
+    #[error("invalid retention, can't be zero or less than window")]
+    InvalidRetention,
+}
+
+/// Slot provides the functionality to set or overwrite the value of any metric
+/// at a specific timestamp.
+#[async_trait]
+pub trait Slot {
+    /// Counter sets (or overrides) the current stored value for this key,
+    /// with value
+    async fn counter(&mut self, key: &str, value: f64) -> Result<()>;
+    /// Key return the key of the slot which is the window timestamp
+    async fn key(&self) -> Result<i64>;
+}
+
+/// RRD is a round robin database of fixed size which is specified on creation
+/// RRD stores `counter` values. Which means values that can only go UP.
+/// then it's easy to compute the increase of this counter over a given window
+/// The database only keep history based on retention.
+#[async_trait]
+pub trait RRD<S, I, 'a>
+where
+    S: Slot,
+    I: Iterator,
+{
+    /// Slot returns the current window (slot) to store values.
+    async fn slot(&'a mut self) -> Result<S>;
+    /// Counters, return all stored counters since the given time (since) until now.
+    async fn counters(&self, since: std::time::SystemTime) -> Result<I>;
+    /// Last returns the last reported value for a metric given the metric
+    /// name
+    async fn last(&self, key: &str) -> Result<Option<f64>>;
+}
+
+/// SqliteRRD is the [`RRD`] implementation using Sqlite under the hood.
+pub struct SqliteRRD {
+    pool: sqlx::Pool<Sqlite>,
+    window: i64,
+    retention: i64,
+}
+
+/// SqliteSlot is the [`Slot`] implementation using Sqlite under the hood.
+pub struct SqliteSlot<'a> {
+    rrd: &'a mut SqliteRRD,
+    key: i64,
+}
+
+struct Counters {
+    index: usize,
+    inner: Vec<Counter>,
+}
+
+pub struct Counter {
+    metric: String,
+    value: f64,
+}
+
+impl Clone for Counter {
+    fn clone(&self) -> Self {
+        Counter {
+            metric: self.metric.clone(),
+            value: self.value,
+        }
+    }
+}
+
+impl Iterator for Counters {
+    type Item = Counter;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index == self.inner.len() {
+            return None;
+        }
+        let ret = self.inner[self.index].clone();
+        self.index += 1;
+        Some(ret)
+    }
+}
+
+impl From<Vec<(String, f64)>> for Counters {
+    fn from(v: Vec<(String, f64)>) -> Self {
+        let mut inner = Vec::new();
+        for r in v {
+            inner.push(Counter {
+                metric: r.0,
+                value: r.1,
+            })
+        }
+        Counters { index: 0, inner }
+    }
+}
```

Comment:
the "right" trait to convert an object to an Iterator is the "IntoIterator" trait from std. You should use this instead (not that you actually need to do manual implementation of the iterator). Why using IntoIterator instead of From (or Into) iterator is that IntoIterator works with the for loop. So if your object implements an IntoIterator you can do this 

```rust
let obj = MyObject{}; // my object implements IntoIterator
for i in obj {
   // work with i.
}
// this code above will call `into_iter()` automatically for you
```